lib/crypto/src/poseidon2/mod.rs

```diff
@@ -0,0 +1,294 @@
+//! This module contains the Poseidon hash ([whitepaper]) function implemented
+//! as a [Sponge Function].
+//!
+//! Poseidon permutation here follows referenced in [whitepaper] original [rust
+//! implementation] with slight improvements.
+//!
+//! [Sponge function]: https://en.wikipedia.org/wiki/Sponge_function
+//! [whitepaper]: https://eprint.iacr.org/2023/323.pdf
+//! [rust implementation]: https://github.com/HorizenLabs/poseidon2
+
+pub mod instance;
+pub mod params;
+
+use alloc::{boxed::Box, vec, vec::Vec};
+
+use crate::{field::prime::PrimeField, poseidon2::params::PoseidonParams};
+
+/// Determines whether poseidon sponge in absorbing or squeezing state.
+/// In squeezing state, sponge can only squeeze elements.
+#[derive(Clone, Copy, Debug)]
+enum Mode {
+    Absorbing,
+    Squeezing,
+}
+
+/// Poseidon2 sponge that can absorb any number of [`F`] field elements and be
+/// squeezed to a finite number of [`F`] field elements.
+#[derive(Clone, Debug)]
+pub struct Poseidon2<P: PoseidonParams<F>, F: PrimeField> {
+    phantom: core::marker::PhantomData<P>,
+    state: Box<[F]>,
+    mode: Mode,
+    index: usize,
+}
+
+impl<P: PoseidonParams<F>, F: PrimeField> Default for Poseidon2<P, F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<P: PoseidonParams<F>, F: PrimeField> Poseidon2<P, F> {
+    /// Create a new Poseidon sponge.
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            phantom: core::marker::PhantomData,
+            state: vec![F::zero(); P::T].into_boxed_slice(),
+            mode: Mode::Absorbing,
+            // Begin index from `CAPACITY`. Skip capacity elements.
+            index: P::CAPACITY,
+        }
+    }
+
+    /// Size of poseidon sponge's state.
+    #[must_use]
+    pub const fn state_size() -> usize {
+        P::T
+    }
+
+    /// Start index of partial rounds.
+    #[must_use]
+    const fn partial_round_start() -> usize {
+        P::ROUNDS_F / 2
+    }
+
+    /// End index of partial rounds (noninclusive).
+    #[must_use]
+    const fn partial_round_end() -> usize {
+        Self::partial_round_start() + P::ROUNDS_P
+    }
+
+    /// Total number of rounds.
+    #[must_use]
+    const fn rounds() -> usize {
+        P::ROUNDS_F + P::ROUNDS_P
+    }
+
+    /// Absorb a single element into the sponge.
+    ///
+    /// # Panics
+    ///
+    /// May panic if absorbing while squeezing.
+    pub fn absorb(&mut self, elem: &F) {
```

Comment:
currently interface of Poseidon looks like that. Within the other pr I'll incorporate it in generic hash interface and add benchmark for it